sql/core/src/main/scala/org/apache/spark/sql/execution/joins/OuterJoin.scala

```diff
@@ -111,51 +124,41 @@ trait HashOuterJoin {
   // iterator for performance purpose.
 
   protected[this] def leftOuterIterator(
-      key: InternalRow,
       joinedRow: JoinedRow,
       rightIter: Iterable[InternalRow]): Iterator[InternalRow] = {
     val ret: Iterable[InternalRow] = {
-      if (!key.anyNull) {
-        val temp = if (rightIter != null) {
-          rightIter.collect {
-            case r if boundCondition(joinedRow.withRight(r)) => resultProjection(joinedRow).copy()
-          }
-        } else {
-          List.empty
-        }
-        if (temp.isEmpty) {
-          resultProjection(joinedRow.withRight(rightNullRow)).copy :: Nil
-        } else {
-          temp
+      val temp = if (rightIter != null) {
+        rightIter.collect {
+          case r if boundCondition(joinedRow.withRight(r)) => resultProjection(joinedRow).copy()
         }
       } else {
+        List.empty
+      }
+      if (temp.isEmpty) {
         resultProjection(joinedRow.withRight(rightNullRow)).copy :: Nil
+      } else {
+        temp
       }
     }
     ret.iterator
   }
 
   protected[this] def rightOuterIterator(
-      key: InternalRow,
       leftIter: Iterable[InternalRow],
       joinedRow: JoinedRow): Iterator[InternalRow] = {
     val ret: Iterable[InternalRow] = {
```

Comment:
Constructing an `Iterable` only to turn around and call `.iterator` doesn't make a lot of sense to me, especially since it forces us to defensively copy in the `collect`.  Instead, I think this should accept and return iterators.